Implement sequence fields that display sequence from genomic coordinates

mdondrup opened the pull request:
<!--- Thank you for contributing! -->
<!--- Provide a general summary of your changes in the Title above -->
<!--- See our Contribution Guidelines here:
          https://tripaldoc.readthedocs.io/en/latest/contributing.html -->


<!---  Please set the header below based on the PR type:
#New Feature
# Bug Fix
# Tripal 4 Core Dev Task  --->

#New Feature

### Issue #1836 #2121
(possibly more)

## Description
<!--- Describe your changes in detail -->
This PR implements sequence fields that compute a sequence for genomic features derived from their genomic location (featureloc in Chado) against their landmark sequence. This new feature eliminates the need to extract feature residues upon import for any genomic feature located on a landmark (chromosome, scaffold, contig).
Features (most already supported by the helper code):

 - extract gene sequences from coordinates and reverse complements if located on reverse strand
 - splices transcripts based on exon information
 - can translate polypeptide sequences based on CDS fragments
 - supports and preserves masking information (capitalization of sequence)
 - supports IUPAC ambiguity codes in revcomp and translation
 - supports all genetic codes including mitchondrial, codes can be inferred from Chado
 - supports multiple alignments of one feature to multiple landmarks (e.g. genome builds)

The interface is still relatively basic and supports only genes. Support for different sequence types and displaying of related and sub-features including coloring, interactive highlighting, graphical display of gene models, are work in progress.

<!--- Why is this change required? What problem does it solve? -->
Handling and displaying sequence data is essential functionality for a genomics database. This PR attempts to lift functionality towards the level that was implemented in Tripal 2.

A more comprehensive discussion is here: https://github.com/tripal/tripal/issues/2121#issuecomment-2681425901

## Testing?
<!--- Please describe in detail how to test these changes. -->

Best to start with an empty container or fresh install.
check out mdondrup:tripal:tv4g1-issue1836-sequence-field-infers-sequence-from-location

 - If running in the tripal container or if the tripal module has been activated on a different branch (even if the genomic fields collection has never been loaded):
run `drush  config:delete tripal.tripalfield_collection.genomic_chado; drush cr` to reload the field collection from yaml
 - Delete the Gene field from tripal data collections if present
 - import Genomic fields collection
 - Create organism (S. cerevisiae) and Assembly (S288c_R64)
 - Import PAU8 GFF file 
 - Publish Tripal gene content
 - Visit PAU8 gene
Genomic sequence is displayed and is identical to the residue sequence

<img width="1145" alt="Screenshot 2025-03-15 at 10 40 08 PM" src="https://github.com/user-attachments/assets/9ea0ffff-2352-495e-9476-fc2e97ea9fdc" />

[PAU8-withgeneseq.txt](https://github.com/user-attachments/files/19270394/PAU8-withgeneseq.txt)

## License
Part of the helper code tripal/tripal_chado/src/Services/GenomicFeatureHelper.php is based on BioPHP by Joseba Bikandi under GPL v2 and  can be used in Tripal under the same license:

// author    Joseba Bikandi
// license   GNU GPL v2
// source code available at  biophp.org

 - https://www.biophp.org/minitools/sequence_manipulation_and_data/
 - https://www.biophp.org/minitools/dna_to_protein




<!--- Reviewers will use this section to test the submission! -->
<!--- If you've implemented PHPUnit tests, you can describe the test cases here. -->

mdondrup committed:
Added GenomicFeatureHelper Service.

mdondrup committed:
First attempt to display an additional field for gene.
 - Added gene_derived_sequence field to tripalfieldcollection
 - Added a static function to fille the field property

This doesn't work though. If it works, it should display "Hello World!""

mdondrup committed:
Needed to fill out base table and column. Now set to residues, which is not ideal.

mdondrup committed:
This shows why the field-callback approach isn't a viable option.
 - I managed to have the function foo as a fallback that gives a message whenever it is
invoked
 - I didn't manage to get the (md5sum) field display the property value (foo)
 - Even if that worked, the functions get invoked way too often:
for example the function gets invoked whenever features are listed like in the
Tripal Content View where the field isn't even displayed.
That would slow this down massively.

mdondrup committed:
First time we get some fields to diplay a value returned by a function (foo, foo1).
Added some Drupal messages to show field callback invocation for sequence types.

Note: The derived sequence shows the same content as the residues field.
TODO: Make my own derived sequence type Plugin class

mdondrup committed:
Added the ChadoAlignedSequenceTypeDefault plugin and modified the tripal genomic field collection accordingly.
 - First time we get to display Hello World in the derived sequence field.
 - Still using the standard sequence formatter widget.

mdondrup committed:
Something is very odd with the parameter parsing
when the function is invoked. The my_record_id parameter cannot be passed.
All related code lines have been marked with TODO

mdondrup committed:
First minimal WORKING example.

To test, import the PAU8 gene GFF and compare residues with the Sequence from Alignment.

They should be identical, except for non-capital base codes from the genome.

The id of the current record is now added to the context array.
The id is used in the sequence helper to retrieve the gene sequence from alignment.

A few warnings from the retrieval code.

mdondrup committed:
Merge branch 'tripal:4.x' into tv4g1-issue1836-sequence-field-infers-sequence-from-location

mdondrup committed:
Update GenomicFeatureHelper.php

Added Author credits and license information

dsenalik commented:
This looks great so far, one thought I have at this early stage is that instead of `ChadoAlignedSequenceTypeDefault` this field be called `ChadoDerivedSequenceTypeDefault`. To me an alignment implies two sequences somehow showing their similarity, and this is more of deriving (either by lookup or through codon translation) one sequence from another. Opinions requested.

And I wanted to help you add the `discover()` function but discovered it is currently difficult because we have never done one for a base table column as with the original sequence field - see Issue #2163

mdondrup commented:
Indeed, the term "alignment" in the genomic sequence context is misleading. I think I was confused by the way this field was labeled in the interface previously. I have been looking at how other databases label this part, and they simply call it "genomic sequence" or just "sequence". Given the new functionality, a gene having separate residues different from the would be the absolute exception, because that would either mean the underlying landmark sequence is wrong or the gene model is wrong and those should be corrected. Of course it might still happen in very rare cases when a gene sequence is re-sequenced, but then again, this sequence should rather be saved as a new unplaced gene. As an example of how the gene display could look, I have had my eyes on SGD, see https://www.yeastgenome.org/locus/S000028510#sequence (selected a gene with introns).

Regarding the discover() functionality, I am not sure I completely follow. The derived sequence field doesn't have a direct source column in any chado table. (I have used "dummy_derived_sequence" because the field definition syntax requires it).

dsenalik commented:
> Regarding the discover() functionality, I am not sure I completely follow. The derived sequence field doesn't have a direct source column in any chado table. (I have used "dummy_derived_sequence" because the field definition syntax requires it).

Well I am thinking the "dummy" column would also work for discover, but first I have to get discover to work for the "regular" sequence field

mdondrup committed:
Changed the name of the field type from "aligned_sequence" to "derived_sequence"

mdondrup committed:
Added own sequence formatter and widget to the derived sequence type.

mdondrup committed:
Merge branch 'tv4g1-issue1836-sequence-field-infers-sequence-from-location' of github.com:mdondrup/tripal into tv4g1-issue1836-sequence-field-infers-sequence-from-location

mdondrup committed:
Changed the label to Derived sequence.

mdondrup committed:
The sequence formatter now adds residue numbers and formats in chunks.
The widget doesn't display anything because the derived sequence cannot be edited.

mdondrup committed:
Merge branch 'tripal:4.x' into tv4g1-issue1836-sequence-field-infers-sequence-from-location

mdondrup commented:
Now the derived sequence type has its own formatted and a dummy widget.

<img width="537" alt="image" src="https://github.com/user-attachments/assets/6963667f-482f-41e2-b3b5-08b3b374e5f4" />


------------------------------------------------------------------------------------------------
Same sequence in SGD:
<img width="970" alt="image" src="https://github.com/user-attachments/assets/7bf96f6c-d044-4526-a0a9-3dfcdcf66fbe" />

Todo: 
- add configuration dropdown
- add download buttons

mdondrup committed:
A little bug in line break calculation

mdondrup committed:
Merge branch 'tv4g1-issue1836-sequence-field-infers-sequence-from-location' of github.com:mdondrup/tripal into tv4g1-issue1836-sequence-field-infers-sequence-from-location

mdondrup committed:
Removed some debug output. Edited some comments.

mdondrup committed:
Added comments and TODO tasks, formatting.

mdondrup committed:
Added a lightweight sequence abstraction classes resembling BioSeq and BioSeqfeature for handling sequence operations.

mdondrup committed:
Further edited documentation.

mdondrup committed:
Merge branch 'tripal:4.x' into tv4g1-issue1836-sequence-field-infers-sequence-from-location
